### PR TITLE
compact raft log smoothly

### PR DESCRIPTION
**Pull request by zhangjinpeng1987**
ref https://github.com/pingcap/tikv/issues/1189

### CODE REVIEW

**Comment by zhangjinpeng1987** [comment]
@siddontang @BusyJay @disksing PTAL

**Comment by siddontang** [comment]
I think we should do more, not only this. And I doubt that this may have little help. 

We may generate many region CompactLog raft commands in one GC tick. The default GC tick interval is 5s, seems a little short.
 
The default GC threshold is 1, means that if leader's current log index is n, and all followers have committed log n - 1, it will send a CompactLog raft command. 

We can think re-use No WAL for compacting log, because RocksDB has already fix this issue. #1106.
We can even consider using single-delete for Raft Log.

**Comment by BusyJay** [comment]
Default threshold is 50. I don't think 5s is too short, if it's under high load, 5s can produce many log entries.

**Comment by zhangjinpeng1987** [comment]
Use no WAL write for compact raft log would just reduce a bit little IO because every raft entry key just contains about 20 bytes, compact 1000 raft log will write about 30KB(add RocksDB internal sequence number, del type) to WAL.

File: src/raftstore/store/worker/compact.rs [hunk 1]
```diff
@@ -141,5 +159,10 @@ impl Runnable<Task> for Runner {
                 }
             }
         }
+
+        if self.need_rest() {
```
Comment (siddontang) [inline src/raftstore/store/worker/compact.rs]: Em, I think we can use if check directly, no need `need_rest`

**Comment by siddontang** [comment]
I suggest using a threshold 1 in one test.

**Comment by zhangjinpeng1987** [comment]
Compact raft log produce just a little IO, the compact speed should not limit currently.

**Pull request closed without merging**